### PR TITLE
Change avatar placeholder from single letter to 2 letters

### PR DESCRIPTION
**Pull request by matchish**
https://github.com/nextcloud/server/issues/18717
Signed-off-by: Sergey Shliakhov <husband.sergey@gmail.com>

### CODE REVIEW

**Comment by matchish** [comment]
@jancborchardt @kesselb

**Comment by matchish** [comment]
I've extracted avatar creating to private methods because I need a user with different display names for each case.

**Review by juliusknorr — APPROVED** [review]

**Comment by juliushaertl** [comment]
Screenshots to make reviews for @nextcloud/designers easier:

image

image

**Review by kesselb — APPROVED** [review]

File: lib/private/Avatar/Avatar.php [hunk 1]
```diff
@@ -89,12 +89,15 @@ public function __construct(ILogger $logger) {
...
 	 * @return string
 	 */
-	private function getAvatarLetter(): string {
+	private function getAvatarText(): string {
 		$displayName = $this->getDisplayName();
 		if (empty($displayName) === true) {
 			return '?';
 		} else {
-			return mb_strtoupper(mb_substr($displayName, 0, 1), 'UTF-8');
+			$firstLetters = array_map(function ($namePart) {
```
Comment (kesselb) [inline lib/private/Avatar/Avatar.php]: To improve readability we could move this out of the else block.

File: lib/private/Avatar/Avatar.php [hunk 2]
```diff
@@ -89,12 +89,15 @@ public function __construct(ILogger $logger) {
...
+				return mb_strtoupper(mb_substr($namePart, 0, 1), 'UTF-8');
+			}, explode(' ', $displayName));
```
Comment (kesselb) [inline lib/private/Avatar/Avatar.php]: It's possible to pass a 3rd argument `$limit` to explode. That makes sure to only return $limit number of array elements (https://3v4l.org/Ns1lt). `array_slice` is not necessary than. Just as note I'm fine with the current approach.

**Comment by kesselb** [comment]
Thanks :+1: Great contribution :tada: 

image

Also seems to work with cyrillic characters. Chinese characters not. That's an issue with the font and also not working on master.

File: lib/private/Avatar/Avatar.php (hunk 1 above)
Comment (matchish) [reply lib/private/Avatar/Avatar.php]: Thanks! Fixed

File: lib/private/Avatar/Avatar.php (hunk 2 above)
Comment (matchish) [reply lib/private/Avatar/Avatar.php]: I agree with you. I've removed `array_slice`

Comment (matchish) [reply lib/private/Avatar/Avatar.php]: I've force-pushed the last commit to reduce polish commits. Sorry about that.

**Comment by skjnldsv** [comment]
> Also seems to work with cyrillic characters. Chinese characters not. That's an issue with the font and also not working on master.

With or without imagick? I think it depends on the fonts installed on the os?

**Comment by kesselb** [comment]
> > Also seems to work with cyrillic characters. Chinese characters not. That's an issue with the font and also not working on master.
> 
> With or without imagick? I think it depends on the fonts installed on the os?

https://github.com/nextcloud/server/blob/3f23ace10b406e9d5cfc5c2086a37ef549d7cbdd/lib/private/Avatar/Avatar.php#L152

I think @jancborchardt discussed this at some other issue. The lack of cyrillic / chinese characters was one motivation to use native fonts. Noto Sans seems to support chinese characters. Probably we have to update the font. But that's a different story ;) If you press merge it's the first pr merged for 19 :tada:

**Comment by matchish** [comment]
@kesselb Seems like some checks failed accidentally. Can't figure out how to re-run it.

**Comment by skjnldsv** [comment]
Tests failing

```json
{
  "reqId": "2PvlYgKwiAD5CzA9a6da",
  "level": 3,
  "time": "2020-01-09T05:27:55+00:00",
  "remoteAddr": "127.0.0.1",
  "user": "admin",
  "app": "no app in context",
  "method": "DELETE",
  "url": "/ocs/v1.php/cloud/users/user1",
  "message": {
    "Exception": "Exception",
    "Message": "Return value of OC\\Avatar\\Avatar::getAvatarText() must be of the type string, null returned",
    "Code": 0,
    "Trace": [
      {
        "file": "/drone/src/lib/private/AppFramework/App.php",
        "line": 125,
        "function": "dispatch",
        "class": "OC\\AppFramework\\Http\\Dispatcher",
        "type": "->",
        "args": [
          {
            "__class__": "OCA\\Provisioning_API\\Controller\\UsersController"
          },
          "deleteUser"
        ]
      },
      {
        "file": "/drone/src/lib/private/AppFramework/Routing/RouteActionHandler.php",
        "line": 47,
        "function": "main",
        "class": "OC\\AppFramework\\App",
        "type": "::",
        "args": [
          "OCA\\Provisioning_API\\Controller\\UsersController",
          "deleteUser",
          {
            "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
          },
          {
            "userId": "user1",
            "_route": "ocs.provisioning_api.Users.deleteUser"
          }
        ]
      },
      {
        "function": "__invoke",
        "class": "OC\\AppFramework\\Routing\\RouteActionHandler",
        "type": "->",
        "args": [
          {
            "userId": "user1",
            "_route": "ocs.provisioning_api.Users.deleteUser"
          }
        ]
      },
      {
        "file": "/drone/src/lib/private/Route/Router.php",
        "line": 299,
        "function": "call_user_func",
        "args": [
          {
            "__class__": "OC\\AppFramework\\Routing\\RouteActionHandler"
          },
          {
            "userId": "user1",
            "_route": "ocs.provisioning_api.Users.deleteUser"
          }
        ]
      },
      {
        "file": "/drone/src/ocs/v1.php",
        "line": 82,
        "function": "match",
        "class": "OC\\Route\\Router",
        "type": "->",
        "args": [
          "/ocsapp/cloud/users/user1"
        ]
      }
    ],
    "File": "/drone/src/lib/private/AppFramework/Http/Dispatcher.php",
    "Line": 109,
    "Previous": {
      "Exception": "TypeError",
      "Message": "Return value of OC\\Avatar\\Avatar::getAvatarText() must be of the type string, null returned",
      "Code": 0,
      "Trace": [
        {
          "file": "/drone/src/lib/private/Avatar/Avatar.php",
          "line": 173,
          "function": "getAvatarText",
          "class": "OC\\Avatar\\Avatar",
          "type": "->",
          "args": []
        },
        {
          "file": "/drone/src/lib/private/Avatar/UserAvatar.php",
          "line": 248,
          "function": "generateAvatar",
          "class": "OC\\Avatar\\Avatar",
          "type": "->",
          "args": [
            "user1",
            1024
          ]
        },
        {
          "file": "/drone/src/lib/private/Avatar/Avatar.php",
          "line": 110,
          "function": "getFile",
          "class": "OC\\Avatar\\UserAvatar",
          "type": "->",
          "args": [
            -1
          ]
        },
        {
          "file": "/drone/src/lib/private/User/User.php",
          "line": 434,
          "function": "get",
          "class": "OC\\Avatar\\Avatar",
          "type": "->",
          "args": [
            -1
          ]
        },
        {
          "file": "/drone/src/apps/dav/lib/CardDAV/Converter.php",
          "line": 148,
          "function": "getAvatarImage",
          "class": "OC\\User\\User",
          "type": "->",
          "args": [
            -1
          ]
        },
        {
          "file": "/drone/src/apps/dav/lib/CardDAV/Converter.php",
          "line": 58,
          "function": "getAvatarImage",
          "class": "OCA\\DAV\\CardDAV\\Converter",
          "type": "->",
          "args": [
            "*** sensitive parameter replaced ***"
          ]
        },
        {
          "file": "/drone/src/apps/dav/lib/CardDAV/SyncService.php",
          "line": 278,
          "function": "createCardFromUser",
          "class": "OCA\\DAV\\CardDAV\\Converter",
          "type": "->",
          "args": [
            "*** sensitive parameter replaced ***"
          ]
        },
        {
          "file": "/drone/src/apps/dav/lib/HookManager.php",
          "line": 148,
          "function": "updateUser",
          "class": "OCA\\DAV\\CardDAV\\SyncService",
          "type": "->",
          "args": [
            "*** sensitive parameters replaced ***"
          ]
        },
        {
          "file": "/drone/src/lib/private/legacy/hook.php",
          "line": 111,
          "function": "changeUser",
          "class": "OCA\\DAV\\HookManager",
          "type": "->",
          "args": [
            {
              "run": true,
              "user": "*** sensitive parameter replaced ***",
              "feature": "*** sensitive parameter replaced ***",
              "value": "*** sensitive parameter replaced ***",
              "old_value": "*** sensitive parameter replaced ***"
            }
          ]
        },
        {
          "file": "/drone/src/lib/private/Server.php",
          "line": 541,
          "function": "emit",
          "class": "OC_Hook",
          "type": "::",
          "args": [
            "OC_User",
            "changeUser",
            {
              "run": true,
              "user": "*** sensitive parameter replaced ***",
              "feature": "*** sensitive parameter replaced ***",
              "value": "*** sensitive parameter replaced ***",
              "old_value": "*** sensitive parameter replaced ***"
            }
          ]
        },
        {
          "function": "OC\\{closure}",
          "class": "OC\\Server",
          "type": "->",
          "args": [
            "*** sensitive parameters replaced ***"
          ]
        },
        {
          "file": "/drone/src/lib/private/Hooks/EmitterTrait.php",
          "line": 100,
          "function": "call_user_func_array",
          "args": [
            {
              "__class__": "Closure"
            },
            [
              "*** sensitive parameter replaced ***",
              "*** sensitive parameter replaced ***",
              "*** sensitive parameter replaced ***",
              "*** sensitive parameter replaced ***"
            ]
          ]
        },
        {
          "file": "/drone/src/lib/private/Hooks/PublicEmitter.php",
          "line": 40,
          "function": "emit",
          "class": "OC\\Hooks\\BasicEmitter",
          "type": "->",
          "args": [
            "\\OC\\User",
            "changeUser",
            [
              "*** sensitive parameter replaced ***",
              "*** sensitive parameter replaced ***",
              "*** sensitive parameter replaced ***",
              "*** sensitive parameter replaced ***"
            ]
          ]
        },
        {
          "file": "/drone/src/lib/private/User/User.php",
          "line": 476,
          "function": "emit",
          "class": "OC\\Hooks\\PublicEmitter",
          "type": "->",
          "args": [
            "\\OC\\User",
            "changeUser",
            [
              "*** sensitive parameter replaced ***",
              "*** sensitive parameter replaced ***",
              "*** sensitive parameter replaced ***",
              "*** sensitive parameter replaced ***"
            ]
          ]
        },
        {
          "file": "/drone/src/lib/private/Avatar/UserAvatar.php",
          "line": 211,
          "function": "triggerChange",
          "class": "OC\\User\\User",
          "type": "->",
          "args": [
            "*** sensitive parameter replaced ***",
            "*** sensitive parameter replaced ***"
          ]
        },
        {
          "file": "/drone/src/lib/private/Server.php",
          "line": 1346,
          "function": "remove",
          "class": "OC\\Avatar\\UserAvatar",
          "type": "->",
          "args": []
        },
        {
          "file": "/drone/src/3rdparty/symfony/event-dispatcher/EventDispatcher.php",
          "line": 260,
          "function": "OC\\{closure}",
          "class": "OC\\Server",
          "type": "->",
          "args": [
            "*** sensitive parameters replaced ***"
          ]
        },
        {
          "file": "/drone/src/3rdparty/symfony/event-dispatcher/EventDispatcher.php",
          "line": 235,
          "function": "doDispatch",
          "class": "Symfony\\Component\\EventDispatcher\\EventDispatcher",
          "type": "->",
          "args": [
            [
              {
                "__class__": "Closure"
              }
            ],
            "*** sensitive parameter replaced ***",
            "*** sensitive parameter replaced ***"
          ]
        },
        {
          "file": "/drone/src/3rdparty/symfony/event-dispatcher/EventDispatcher.php",
          "line": 73,
          "function": "callListeners",
          "class": "Symfony\\Component\\EventDispatcher\\EventDispatcher",
          "type": "->",
          "args": [
            [
              {
                "__class__": "Closure"
              }
            ],
            "*** sensitive parameter replaced ***",
            "*** sensitive parameter replaced ***"
          ]
        },
        {
          "file": "/drone/src/lib/private/EventDispatcher/SymfonyAdapter.php",
          "line": 70,
          "function": "dispatch",
          "class": "Symfony\\Component\\EventDispatcher\\EventDispatcher",
          "type": "->",
          "args": [
            "*** sensitive parameter replaced ***",
            "*** sensitive parameter replaced ***"
          ]
        },
        {
          "file": "/drone/src/lib/private/User/User.php",
          "line": 203,
          "function": "dispatch",
          "class": "OC\\EventDispatcher\\SymfonyAdapter",
          "type": "->",
          "args": [
            "*** sensitive parameter replaced ***",
            "*** sensitive parameter replaced ***"
          ]
        },
        {
          "file": "/drone/src/apps/provisioning_api/lib/Controller/UsersController.php",
          "line": 658,
          "function": "delete",
          "class": "OC\\User\\User",
          "type": "->",
          "args": []
        },
        {
          "file": "/drone/src/lib/private/AppFramework/Http/Dispatcher.php",
          "line": 170,
          "function": "deleteUser",
          "class": "OCA\\Provisioning_API\\Controller\\UsersController",
          "type": "->",
          "args": [
            "user1"
          ]
        },
        {
          "file": "/drone/src/lib/private/AppFramework/Http/Dispatcher.php",
          "line": 99,
          "function": "executeController",
          "class": "OC\\AppFramework\\Http\\Dispatcher",
          "type": "->",
          "args": [
            {
              "__class__": "OCA\\Provisioning_API\\Controller\\UsersController"
            },
            "deleteUser"
          ]
        },
        {
          "file": "/drone/src/lib/private/AppFramework/App.php",
          "line": 125,
          "function": "dispatch",
          "class": "OC\\AppFramework\\Http\\Dispatcher",
          "type": "->",
          "args": [
            {
              "__class__": "OCA\\Provisioning_API\\Controller\\UsersController"
            },
            "deleteUser"
          ]
        },
        {
          "file": "/drone/src/lib/private/AppFramework/Routing/RouteActionHandler.php",
          "line": 47,
          "function": "main",
          "class": "OC\\AppFramework\\App",
          "type": "::",
          "args": [
            "OCA\\Provisioning_API\\Controller\\UsersController",
            "deleteUser",
            {
              "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
            },
            {
              "userId": "user1",
              "_route": "ocs.provisioning_api.Users.deleteUser"
            }
          ]
        },
        {
          "function": "__invoke",
          "class": "OC\\AppFramework\\Routing\\RouteActionHandler",
          "type": "->",
          "args": [
            {
              "userId": "user1",
              "_route": "ocs.provisioning_api.Users.deleteUser"
            }
          ]
        },
        {
          "file": "/drone/src/lib/private/Route/Router.php",
          "line": 299,
          "function": "call_user_func",
          "args": [
            {
              "__class__": "OC\\AppFramework\\Routing\\RouteActionHandler"
            },
            {
              "userId": "user1",
              "_route": "ocs.provisioning_api.Users.deleteUser"
            }
          ]
        },
        {
          "file": "/drone/src/ocs/v1.php",
          "line": 82,
          "function": "match",
          "class": "OC\\Route\\Router",
          "type": "->",
          "args": [
            "/ocsapp/cloud/users/user1"
          ]
        }
      ],
      "File": "/drone/src/lib/private/Avatar/Avatar.php",
      "Line": 100
    },
    "CustomMessage": "--"
  },
  "userAgent": "GuzzleHttp/6.3.3 curl/7.38.0 PHP/7.3.12-1+0~20191128.49+debian8~1.gbp24559b",
  "version": "18.0.0.8"
}
```

**Comment by matchish** [comment]
@skjnldsv I've fixed typo but don't understand why a build is failing now. I run tests on a local machine and its passes.

**Comment by skjnldsv** [comment]
> @skjnldsv I've fixed typo but don't understand why a build is failing now. I run tests on a local machine and its passes.

sometimes tests can timeout or crash :)
Let's restart! This si for nc 19 anyway, there is no rush :wink:

**Comment by matchish** [comment]
@skjnldsv Restart button inactive for me. Could you restart build? Thanks)

**Review by jancborchardt — APPROVED** [review]
Super nice design wise :+1: great job @matchish!

**Comment by matchish** [comment]
> Super nice design wise  great job @matchish!

Thanks) I'm happy to be part of this great community!

Could you restart `continuous-integration/drone/pr` it fails again?

**Comment by juliushaertl** [comment]
Remaining test failures seem unrelated

**Comment by matchish** [comment]
@juliushaertl You're right. I've restarted tests ten times and tests fail always in different places.

**Comment by jancborchardt** [comment]
So do we merge this? @skjnldsv want to give it a last look? :)

**Comment by skjnldsv** [comment]
@matchish please rebase and squash your commits in 1 (or allow us to push to your branch ;) )

**Comment by rullzer** [comment]
seems this merge broke the unit tests...

**Comment by nickvergessen** [comment]
I think this needs a bit more fixing. Escpecially in mentions it looks a bit awkward now, as it is not readable at all:
Bildschirmfoto von 2020-01-31 16-45-31

Any ideas what we could do there? Showing a single character in small and 2 in big is not going to work.

**Comment by raimund-schluessler** [comment]
Is it really necessary to show the auto-generated avatar in mentions? It doesn't really add information in my opinion as the user name is written next to it anyway.

**Comment by nickvergessen** [comment]
It helps visually to find the mentions e.g. when scrolling through a long chat history, anyway this is not the right place to discuss this.

**Comment by nickvergessen** [comment]
Also core/src/jquery/avatar.js was not updated and therefor now produces a different avatar then whatever the other method is.

**Pull request merged**